flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala

```diff
@@ -100,18 +96,32 @@ object TableSourceUtil {
           throw new ValidationException(s"Rowtime field '$name' has invalid type $t. " +
             s"Rowtime attributes must be of TimestampType.")
         }
-        val (physicalName, idx, tpe) = resolveInputField(name, tableSource)
+        val (physicalName, idx, logicalType) = resolveInputField(name, tableSource)
         // validate that mapped fields are are same type
-        if (!isAssignable(fromTypeInfoToLogicalType(tpe), t)) {
+        if (!isAssignable(logicalType, t)) {
           throw new ValidationException(s"Type $t of table field '$name' does not " +
-            s"match with type $tpe of the field '$physicalName' of the TableSource return type.")
+            s"match with type $logicalType of the field '$physicalName' of the " +
+            "TableSource return type.")
+        } else if (!isInteroperable(logicalType, t)) {
```

Comment:
Source sink maybe can unify? Use same check?

Reply:
I though about this. But I think we can loosen the restriction for sinks, because 
1) The conversion from internal type (Deciaml) to external type (BigDecimal) doesn't loose precision, because the internal type is already match to the expected type defined in DDL.
2) it's pretty hard to give a correct consumed DataType for `UpsertStreamTableSink`.

Reply:
I think you should add comments to both source and sink to explain why they are different.